### DISCUSSION TITLE
Choose an output device when play a audio file

### DISCUSSION THREAD

**Discussion by pedrolopix** [FactoryXCode/MfPack#68] [Q&A] [answered]
How choose an output audio device to play the audio file?
Also, how to enumerate all output available?

**Comment by FactoryXCode** [comment]
Hi.
The endpoint (output device) will always be the "default endpoint" when playing sound. 
The default endpoint is set by the user in the Windows operating system.
This can be speakers or headphone or whatever output (endpoint) the soundcard has.
Only one of these can be set when having one soundcard.
So, it's not possible to render audio from different sources to headphones and speakers on one soundcard.
If you want to do so, you need at least 2 soundcards, for example play system sound messages to one and music to the other one.

Now the other thing:
Getting the rendering or capturing devices (endpoints) on your system can be called with method GetEndpointDevices declared in WinApi.CoreAudioApi.MMDevApiUtils.pas.
This unit contains most common API methods / utilities for the core audio API.

Regards, Tony.

**Reply by FactoryXCode** [reply, marked as answer]
Also, You can not change the system output devices from within your software system wide. Until Win 10 this could be, but in win 11, that's not possible anymore. see update KB5030509 (KB5030509’s enhanced mixer explicitly mentions per-app customization and device swapping in the UI), the way audio output is managed, introducing the enhanced volume mixer in Quick Settings.
In short, you could do that, but per app, not system wide, because there is no guarantee MS will change that one again. And side effects are: when different app are using the same default output device. Now, the trick is, and that works also on the latest Windows 11, explicitly assign a renderer to an audio endpoint, (you should do that by using the DeviceID (which looks like {0.0.0.00000000}.{d7e13618-ca94-4b30-a916-0934986710a9} ), store that in a setupfile or something, so the next time you start your app, it will not fallback to the system default. A note of precaution: When possible leave the system Default endpoint untouched, only use other active endpoint devices.  We are working on a sample, on how to do that.
